core/block.js

```diff
@@ -1424,16 +1424,19 @@ Blockly.Block.prototype.toString = function(opt_maxLength, opt_emptyToken) {
 
   // Run through our text array and simplify expression to remove parentheses
   // around single field blocks.
-  for (var i = 2, l = text.length; i < l; i++) {
+  // E.g. ['repeat', '(', '10', ')', 'times', 'do', '?']
+  for (var i = 2; i < text.length; i++) {
     if (text[i - 2] == '(' && text[i] == ')') {
       text[i - 2] = text[i - 1];
       text.splice(i - 1, 2);
-      l -= 2;
     }
   }
 
-  // Join the text array, removing spaces around added paranthesis.
-  text = text.join(' ').replace(/(\() | (\))/gmi, '$1$2').trim() || '???';
+  // Join the text array, removing spaces around added parentheses.
+  text = text.reduce(function(acc, value) {
```

Comment:
Are we allowed to use reduce in core? If so, I am super excited about that :D

Reply:
I don't see why not.  Supported in IE9.  Already used in Blockly.utils.string.shortestStringLength